modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterGroupImpl.java

```diff
@@ -17,51 +17,759 @@
 
 package org.apache.ignite.internal.client.thin;
 
+import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 import org.apache.ignite.client.ClientClusterGroup;
+import org.apache.ignite.client.ClientException;
+import org.apache.ignite.client.ClientFeatureNotSupportedByServerException;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.binary.BinaryReaderExImpl;
+import org.apache.ignite.internal.binary.BinaryWriterExImpl;
 import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.A;
 import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteProductVersion;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.IgniteNodeAttributes.ATTR_MACS;
 
 /**
  * Implementation of {@link ClientClusterGroup}.
  */
-@SuppressWarnings("AssignmentOrReturnOfFieldWithMutableType")
 class ClientClusterGroupImpl implements ClientClusterGroup {
-    /** Node id's. */
-    private final Collection<UUID> nodeIds;
+    /** Channel. */
+    protected final ReliableChannel ch;
+
+    /** Binary marshaller. */
+    protected final ClientBinaryMarshaller marsh;
+
+    /** Projection filters. */
+    private final ProjectionFilters projectionFilters;
+
+    /** Cached topology version. */
+    private long cachedTopVer;
```

Comment:
It is just my opinion, but at fisrt look, it can be unclear, that reading this fields (`cachedTopVer` and `cachedNodeIds`) is unsafe without obtaining lock over instance of `ClientClusterGroupImpl`. This should be emphasized.

Reply:
Added comment to javadoc